drivers/disk/flashdisk.c

```diff
@@ -96,36 +150,27 @@ static int disk_flash_access_read(struct disk_info *disk, uint8_t *buff,
 }
 
 /* This performs read-copy into an output buffer */
-static int read_copy_flash_block(off_t start_addr, uint32_t size,
+static int read_copy_flash_block(struct disk_info *disk,
+				 off_t start_addr,
+				 uint32_t size,
 				 const void *src_buff,
 				 uint8_t *dest_buff)
 {
+	struct flashdisk_data *ctx;
 	off_t fl_addr;
-	uint32_t num_read;
-	uint32_t offset = 0U;
+	uint32_t offset;
+
+	ctx = CONTAINER_OF(disk, struct flashdisk_data, info);
 
 	/* adjust offset if starting address is not erase-aligned address */
-	if (start_addr & (CONFIG_DISK_FLASH_ERASE_ALIGNMENT - 1)) {
-		offset = start_addr & (CONFIG_DISK_FLASH_ERASE_ALIGNMENT - 1);
-	}
+	offset = start_addr & (ctx->page_size - 1);
 
 	/* align starting address to an aligned address for flash erase-write */
-	fl_addr = ROUND_DOWN(start_addr, CONFIG_DISK_FLASH_ERASE_ALIGNMENT);
-
-	num_read = GET_NUM_BLOCK(CONFIG_DISK_ERASE_BLOCK_SIZE,
-				 CONFIG_DISK_FLASH_MAX_RW_SIZE);
+	fl_addr = ROUND_DOWN(start_addr, ctx->page_size);
```

Comment:
What do you mean? There is no check here. This function:
- reads out whole page from the underlying flash memory (therefore it rounds down to `ctx->page_size`)
- overwrites the read data with user supplied data, so `dest_buff` holds the page as it should appear on disk after write

This code reads more than strictly necessary as the "overwritten data part" does not need to be read from flash. It could be optimized to skip reading the overwritten data by having two `flash_read()` calls and some local variables.

Reply:
Yeah, sorry. What I meant is that the `read_copy_flash_block` should be called with `start_addr` already rounded down properly. I have incorrectly read the `update_flash_block` and failed to notice that `fl_addr` there is not used with `read_copy_flash_block`, but with `overwrite_flash_block`.